lager/detail/smart_lens.hpp

```diff
@@ -46,9 +58,32 @@ struct smart_lens_base
  * Try to find a nice lens type for the give type.
  */
 template <typename T>
-struct smart_lens : smart_lens_base
+struct smart_lens : smart_lens_base<T>
 {};
 
+struct functor_wrapper_t {
+    template <typename T>
+    const_functor<T> operator()(T&& t) const noexcept;
+};
+
+template <typename Lens, typename T>
+using viewed_t = decltype(
+    std::declval<Lens>()(functor_wrapper_t{})(std::declval<T>()).value);
```

Comment:
Why not: `decltype(view(declval<Lens>(), declval<T>))`?

Reply:
same as above, can't sfinae on lenses...